Support for Audio Dataset Loading and Processing

Helw150 opened the pull request:
Wired everything up to get to a loader that seems functional (at least for the purposes I have for it). Definitely lot's of room for discussion here, since some parts definitely didn't have straightforward corollaries in the text data. 

One thing that I definitely want thoughts on: The existing dataloaders mostly operate under the assumption that data can be packed together - which makes sense for LM pretraining. This seems less straightforward for Multimodal Encoder-Decoder setups like whisper (since there's not a clear prefix vs. suffix). Here, I've opted to pad everything to the tokenizer max_length. Similarly, while data is batched together during processing, I flatten it back out.

One other thing: This code repeats a fair amount of stuff from the `text` datasets. This code could be a lot more DRY but that would require some Mixins that could be re-used across both this and `text`. Happy to work those in to this PR if it's desirable, but opted to repeat where it was non-trivial to compose with existing classes for at least the first draft!

Helw150 commented:
Looks like I'll need to create a small test dataset on HuggingFace. The tests here are passing locally, but looks like they get killed in CI since it's processing a real dataset, rather than a toy.

dlwh commented on src/levanter/data/audio.py:
```diff
@@ -0,0 +1,512 @@
+                attn_mask = hax.named(inputs["attention_mask"], self.KPos)
+                attn_mask = AttentionMask.explicit(attn_mask)
+
+                if self.fcm_prob > 0:
```
probably just rip this out for whisper unless you really want it. would be good to dry this (but probably we want to change where this goes...)

dlwh commented on src/levanter/data/sharded_dataset.py:
```diff
@@ -228,6 +229,60 @@ def open_shard_at_row(self, shard_name: str, row: int) -> Iterator[str]:
+    def shard_names(self) -> Sequence[str]:
+        return list(self._shard_name_to_url_mapping.keys())
+
+    def _resolve_audio_pointer(self, audio_pointer) -> dict[str, Any]:
```
should this not support "bytes" as well?

dlwh reviewed (approved):
definitely agree there's a lot of duplication. I'm not against repetition in all cases but does feel like there are some utilities we could extract. but it's not pressing. rule of three and all that

Helw150 commented on src/levanter/data/audio.py:
```diff
@@ -0,0 +1,512 @@
+                text = x[self.text_key]
+                audio = x[self.audio_key]
+                if "array" in audio:
+                    return (audio["array"], audio["sampling_rate"], text)
```
This is the HuggingFace "Audio" format (it's just a dict under the hood). I refactored to unify this all in the AudioTextUrlDataset processor with a pointer to the docs about this data type!

Helw150 commented on src/levanter/compat/hf_checkpoints.py:
```diff
@@ -810,6 +815,28 @@ def load_tokenizer(model_name_or_path, revision=None, local_cache_dir=None, trus
 
 
+def load_processor(model_name_or_path, revision=None, local_cache_dir=None, trust_remote_code=True) -> ProcessorMixin:
+    """Like AutoTokenizer.from_pretrained, but works with gs:// paths or anything on fsspec"""
```
Moved it into a function that takes an arbitrary from_pretrained method!

dlwh replied:
thank you!

Helw150 commented on src/levanter/data/audio.py:
```diff
@@ -0,0 +1,512 @@
+
+logger = logging.getLogger("levanter.data.audio")
+
+AudioTextStorageBatch = TypedDict(
```
Admittedly, the level of type-checking MyPy does with this seems minimal :laughing:

dlwh replied:
lol, typical

Helw150 commented:
Addressed PR Comments! Leaving unmerged for now in-case of follow-up comments on the changes. Will merge tomorrow afternoon otherwise.

I made a Parquet Version of HuggingFace's testing audio dataset to make it work with streaming: https://huggingface.co/datasets/WillHeld/test_librispeech_parquet . If you'd rather have all the testing data hosted on a CRFM HuggingFace account, let me know!

Next steps are to implement a trainer that puts all the previous pieces together to actually train a Whisper model. I'll pattern off one of the existing tutorials so that I can write up a quick tutorial so bring the Levanter sales pitch to a new modality!

dlwh commented on src/levanter/data/shard_cache.py:
```diff
@@ -1737,6 +1737,10 @@ def unshard(self):
         """
         return ShardCache(self.cache_dir, self._batch_size, self._ledger, self._broker, 0, 1)
 
+    def set_batch_size(self, batch_size):
```
i strongly prefer immutability by default (with an exception for iterators and performance). could we do with_batch_size and reutrn a copy?

Helw150 replied:
Yes - of course! As a former Scala dev, I should have defaulted to the immutable version :laughing:

dlwh replied:
Now it makes sense why you figured all this out so quickly :)

dlwh reviewed (commented):
Please remove the mutability from ShardCache, then ok to merge.

Helw150 commented:
Waiting for tests to past, then merging!
